### PR TITLE
[MRG] Make events files fetched via dataset fetchers BIDS compliant

### PR DESCRIPTION
**Pull request by kchawla-pi**
issue #234 
First-level localizer fMRI dataset events file is now BIDS compliant

### CODE REVIEW

**Comment by bthirion** [comment]
Please also consider the other diowloaded files (spm_auditory, spm_multimodal_faces).
Thx.

**Comment by kchawla-pi** [comment]
Yeah, I figured I'll work first on https://github.com/nistats/nistats/issues/223#issuecomment-426573423
. You want to wrap this up first?

**Comment by bthirion** [comment]
To me these things go together. You can handle them in the same PR if it's easier.

**Comment by kchawla-pi** [comment]
From now on, and in all examples, while reading events files, we'll either have to use `pandas.read_csv(sep='\t')` or `pandas.read_table()` .
We should check and modify the examples accordingly.

**Comment by kchawla-pi** [comment]
@bthirion 2nd fetcher landed. Ready for review.

File: examples/01_tutorials/plot_single_subject_single_run.py [hunk 1]
```diff
@@ -65,38 +65,13 @@
...
-    42  ,  504  ,  rest
-    42  ,  546  ,  active
-    42  ,  588  ,  rest
-    42  ,  630  ,  active
-"""
-
-###############################################################################
-# We can read such a table from a spreadsheet file  created with OpenOffice Calcor Office Excel, and saved under the *comma separated values* format (``.csv``). 
+# timing of the auditory stimulation and rest periods. This is typically
+# provided in an events.tsv or events.csv file. The path of this file is
```
Comment (bthirion) [inline examples/01_tutorials/plot_single_subject_single_run.py]: I'd simply put: "events.tsv file"

File: examples/01_tutorials/plot_single_subject_single_run.py [hunk 2]
```diff
@@ -65,38 +65,13 @@
...
+# provided in the dataset.
+# Please note that events are sometimes called paradigms, though this
```
Comment (bthirion) [inline examples/01_tutorials/plot_single_subject_single_run.py]: I'd remove that statement. I don't think we should explicitly argue about terminology

File: nistats/tests/test_make_bids_compliant_localizer_first_level_paradigm_file.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import os
```
Comment (bthirion) [inline nistats/tests/test_make_bids_compliant_localizer_first_level_paradigm_file.py]: I'm wondering why you're not adding that one in test_datasets.py directly. I think I wouldn't like to see 100 files in this directory.

File: nistats/tests/test_make_spm_auditory_events_file.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import os
```
Comment (bthirion) [inline nistats/tests/test_make_spm_auditory_events_file.py]: Same thing here.

**Comment by bthirion** [comment]
Let me know when this is ready for review.

**Comment by kchawla-pi** [comment]
Absolutely. 
Currently investigating a weird glitch where the tests are failing because the .mat files are empty during nosetests, but are fine when I run the fetch_spm_multimodal_fmri() .

**Comment by kchawla-pi** [comment]
It seems the failure is due to the way the test is written. Dummy .mat files are created (which are empty) during the tests and this was fine because those .mat files were not being loaded during testing.
However, when the events file are being generated in the fetcher itself, that requires the .mat files to be loaded using `loadmat()`. Since they are empty, the tests are failing.

**Comment by kchawla-pi** [comment]
@bthirion 
For now I have programmed it to generate a warning during nosetests about the inability to generate events files when the Mat files are empty.
A solution would be to download the dataset from the server when this happens, instead of the mock data currently used. However that will add 2 minutes of download time everytime the tests are run locally.
I haven't written a unit test for this yet, other than that, this is ready for review.

File: nistats/datasets.py [hunk 5]
```diff
@@ -334,69 +485,134 @@ def fetch_spm_auditory(data_dir=None, data_name='spm_auditory',
...
+        if not _subject_data:
+            return None
+        try:
+            paradigm = _make_events_file_spm_multimodal_fmri(_subject_data, session)
+        except MatReadError as mat_err:
+            warnings.warn('{}. An events.tsv file cannot be generated'.format(str(mat_err)))
+        else:
+            events_filepath = _make_events_filepath_spm_multimodal_fmri(_subject_data, session)
+            paradigm.to_csv(events_filepath, sep='\t', index=False)
+            _subject_data['events{}'.format(session+1)] = events_filepath
```
Comment (bthirion) [inline nistats/datasets.py]: pep8 'session + 1'

File: nistats/datasets.py [hunk 6]
```diff
@@ -334,69 +485,134 @@ def fetch_spm_auditory(data_dir=None, data_name='spm_auditory',
...
+    return _glob_spm_multimodal_fmri_data(subject_dir)
 
-        return Bunch(**_subject_data)
 
-    # maybe data_dir already contains the data ?
-    data = _glob_spm_auditory_data()
-    if data is not None:
-        return data
+def _make_events_filepath_spm_multimodal_fmri(_subject_data, session):
+    key = 'trials_ses{}'.format(session+1)
```
Comment (bthirion) [inline nistats/datasets.py]: same comment

File: nistats/datasets.py [hunk 7]
```diff
@@ -334,69 +485,134 @@ def fetch_spm_auditory(data_dir=None, data_name='spm_auditory',
...
+    events_file_location = os.path.dirname(_subject_data[key])
+    events_filename = 'session{}_events.tsv'.format(session+1)
```
Comment (bthirion) [inline nistats/datasets.py]: again

File: nistats/datasets.py (hunk 5 above)
Comment (kchawla-pi) [reply nistats/datasets.py]: In standalone expressions, not in function/method args.

File: nistats/datasets.py (hunk 6 above)
Comment (kchawla-pi) [reply nistats/datasets.py]: In standalone expressions, not in function/method args

File: nistats/datasets.py (hunk 7 above)
Comment (kchawla-pi) [reply nistats/datasets.py]: In standalone expressions, not in function/method args

Comment (kchawla-pi) [reply nistats/datasets.py]: Regardless, I'll obviate this concern soon

File: nistats/datasets.py (hunk 6 above)
Comment (kchawla-pi) [reply nistats/datasets.py]: Regardless, I'll obviate this concern soon

File: nistats/datasets.py (hunk 5 above)
Comment (kchawla-pi) [reply nistats/datasets.py]: Regardless, I'll obviate this concern soon

**Comment by kchawla-pi** [comment]
I can't write the missing unit test withut bundling a 1.2KB .mat file, hence that will keep.
One final review and this is done AFAIK.

**Comment by kchawla-pi** [comment]
Make sure to update the nilearn_data using the new fetchers for the tutorial. Else the new code will not work.

**Comment by bthirion** [comment]
It's fine to ship a 1.2K mat file with the library.

**Comment by kchawla-pi** [comment]
Turns out it is more complex than that. I don't think we need to focus on this right away, post release, I will add tests.

**Comment by bthirion** [comment]
Can you rebase on master ?

**Comment by bthirion** [comment]
Great ! I'll look at that tomorrow.

**Comment by bthirion** [comment]
Great job, merging.

**Comment by kchawla-pi** [comment]
Cool, I'll get started on the events not paradigms PR.

**Pull request merged**